rqd/rqd/rqcore.py

```diff
@@ -1219,6 +1219,8 @@ def print_and_flush_ln(fd, last_timestamp):
 
         remainder = lines[-1]
         for line in lines[0:-1]:
+            # Convert to ASCII while discarding characters that can not be encoded
+            line = line.encode('ascii', 'ignore')
```

Comment:
@ramonfigueiredo Thank you for the PR and the detailed write up. The detail makes it nice and easy to review with the proper context.

So, this error occurs when we intercept output in order to prepend a timestamp. What is logged to the file when `RQD_PREPEND_TIMESTAMP` is `False`?

My sense is that the output which is logged when `RQD_PREPEND_TIMESTAMP` is `True` vs `False` should be as similar as possible, aside from the timestamp of course.

Reply:
Dear @bcipriano ,

Sorry for the delay in answering your question. Thank you for your thorough review and the positive feedback on the PR and write-up. 

Regarding your comment, the `UnicodeEncodeError` occurs when we intercept output to prepend a timestamp. When `RQD_PREPEND_TIMESTAMP` is `False`, the output is logged directly without any modification or encoding to ASCII, preserving the original characters, including any non-ASCII ones.

Here is how the output differs based on the value of RQD_PREPEND_TIMESTAMP:

1. When `RQD_PREPEND_TIMESTAMP` is `True`:
- We intercept the output to prepend a timestamp.
- Non-ASCII characters are encoded to ASCII with the 'ignore' option, discarding any non-ASCII characters.
- Example log: "[12:34:56] text here , caf" (non-ASCII characters are ignored).

2. When `RQD_PREPEND_TIMESTAMP` is `False`:

- The output is logged directly without any modification.
- Non-ASCII characters are preserved.
- Example log: "text here 영화, café" (non-ASCII characters are retained).

To ensure the outputs are as similar as possible, aside from the timestamp, I can adjust our approach to handle encoding more gracefully. Instead of discarding non-ASCII characters, I can consider other strategies, such as escaping them or converting them to a specific placeholder.

However, given the current solution, the primary goal was to prevent crashes due to `UnicodeEncodeError` by ignoring non-ASCII characters when `RQD_PREPEND_TIMESTAMP` is `True` (default option). This approach was chosen for simplicity and robustness.

If maintaining non-ASCII characters is critical, I can explore additional strategies to handle encoding more gracefully without discarding valuable information. I am open to suggestions and further discussion on the best approach to balance robustness and information retention.

Thank you again for your review and insights. I look forward to your feedback.

Reply:
Hi @bcipriano 

FYI ...

The updated logic to ensure consistent handling of non-ASCII characters in logs with RQD_PREPEND_TIMESTAMP = True or RQD_PREPEND_TIMESTAMP = False.